File: spotlight-sarajevo-be/src/main/java/com/spotlightsarajevo/modules/transport/service/TransportService.java
package com.spotlightsarajevo.modules.transport.service;

import com.spotlightsarajevo.modules.transport.api.dto.TransportMethodLineModel;
import com.spotlightsarajevo.modules.transport.api.dto.TransportMethodModel;
import com.spotlightsarajevo.modules.transport.api.dto.TransportMethodShorthandModel;
import com.spotlightsarajevo.modules.transport.api.dto.TransportTaxiOperatorModel;
import org.springframework.http.ResponseEntity;

import java.util.List;

/**
 * TransportService interface defines methods for retrieving transport-related data.
 * It includes methods to find all transport methods, find a transport method by its ID,
 * find transport lines by operator and transport type, and find taxi operators.
 */
public interface TransportService {
    /**
     * Retrieves a list of all transport methods in shorthand format.
     *
     * @return ResponseEntity containing a list of TransportMethodShorthandModel
     */
    ResponseEntity<List<TransportMethodShorthandModel>> findAllTransportMethods();

    /**
     * Retrieves a transport method by its unique identifier.
     *
     * @param id the unique identifier of the transport method
     * @return ResponseEntity containing the TransportMethodModel
     */
    ResponseEntity<TransportMethodModel> findTransportMethodById(Integer id);

    /**
     * Retrieves transport lines based on the operator ID and transport type ID.
     *
     * @param operatorId the unique identifier of the transport operator
     * @param transportTypeId the unique identifier of the transport type
     * @return ResponseEntity containing a list of TransportMethodLineModel
     */
    ResponseEntity<List<TransportMethodLineModel>> findLinesByOperatorAndTransportType(Integer operatorId, Integer transportTypeId);

    /**
     * Retrieves a list of all taxi operators.
     *
     * @return ResponseEntity containing a list of TransportTaxiOperatorModel
     */
    ResponseEntity<List<TransportTaxiOperatorModel>> findTaxiOperators();
}

File: spotlight-sarajevo-fe/src/app/services/transport.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../environments/environment';
import { HttpClient } from '@angular/common/http';
import {
  TaxiOperatorModel,
  TransportMethodLineModel,
  TransportMethodModel,
  TransportMethodShorthandModel,
} from '../shared/models/transport.model';

/**
 * Service for interacting with the Public Transport API.
 * Provides methods to fetch transport methods, lines, and taxi operators.
 * All requests include credentials for authentication.
 * 
 * Methods and entities incorporated in the service: TransportMethodShorthandModel, TransportMethodModel, TransportMethodLineModel, TaxiOperatorModel
 * 
 * @version 1.0.0
 * @author hajrudin.imamovic
 */
@Injectable({
  providedIn: 'root',
})
export class PublicTransportService {
  private API_URL = environment.API_URL;

  constructor(private http: HttpClient) {}

  /**
   * Fetches all transport methods in shorthand format.
   * @returns An observable of an array of TransportMethodShorthandModel.
   */
  findAllTransportMethods() {
    return this.http.get<TransportMethodShorthandModel[]>(this.API_URL + `/transport/all`, {
      withCredentials: true,
    });
  }

  /**
   * Fetches detailed information about a specific transport method by its ID.
   * @param id - The ID of the transport method to fetch.
   * @returns An observable of TransportMethodModel.
   */
  findMethodById(id: number) {
    return this.http.get<TransportMethodModel>(this.API_URL + `/transport/${id}`, {
      withCredentials: true,
    });
  }

  /**
   * Fetches transport lines for a specific operator and transport type.
   * @param operator - The ID of the transport operator.
   * @param transportType - The type of transport (e.g., tramcar, trolley).
   * @returns An observable of an array of TransportMethodLineModel.
   */
  findLinesByOperatorAndTransportType(operator: number, transportType: number) {
    return this.http.get<TransportMethodLineModel[]>(
      this.API_URL + `/transport/lines?operatorId=${operator}&transportTypeId=${transportType}`,
      {
        withCredentials: true,
      },
    );
  }

  /**
   * Fetches all taxi operators.
   * @returns An observable of an array of TaxiOperatorModel.
   */
  findTaxiOperators() {
    return this.http.get<TaxiOperatorModel[]>(this.API_URL + `/transport/taxis`, {
      withCredentials: true,
    });
  }
}
